redfish/collector.go

```diff
@@ -0,0 +1,241 @@
+package redfish
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/setup-hw/config"
+	"github.com/cybozu-go/setup-hw/gabs"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const namespace = "hw"
+
+// Collector implements prometheus.Collector interface.
+type Collector struct {
+	rules  []convertRule
+	client *client
+	cache  *cache
+}
+
+// CollectorConfig is a set of configurations for Collector.
+type CollectorConfig struct {
+	AddressConfig *config.AddressConfig
+	Port          string
+	UserConfig    *config.UserConfig
+	Rule          []byte
+}
+
+// NewCollector returns a new instance of Collector.
+func NewCollector(cc *CollectorConfig) (*Collector, error) {
+	cache := &cache{dataMap: make(dataMap)}
+	client, err := newClient(cc, cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var rules []convertRule
+	err = yaml.Unmarshal(cc.Rule, &rules)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rule := range rules {
+		if err := rule.Validate(); err != nil {
+			return nil, err
+		}
+	}
+
+	return &Collector{rules: rules, client: client, cache: cache}, nil
+}
+
+// Describe does nothing for now.
+func (c Collector) Describe(ch chan<- *prometheus.Desc) {
+}
+
+// Collect sends metrics collected from BMC via Redfish.
+func (c Collector) Collect(ch chan<- prometheus.Metric) {
+	dataMap := c.cache.get()
+
+	for _, rule := range c.rules {
+		for path, parsedJSON := range dataMap {
+			matched, pathLabels := matchPath(rule.Path, path)
+			if !matched {
+				continue
+			}
+
+			for _, propertyRule := range rule.Rules {
+				matchedProperties := matchPointer(propertyRule.Pointer, parsedJSON, path)
+				for _, matched := range matchedProperties {
+					value, err := propertyRule.Converter(matched.property)
+					if err != nil {
+						log.Warn("failed to interpret Redfish data as metric", map[string]interface{}{
```

Comment:
Can this log become too frequent?
If so, please suppress too many logs here.

Reply:
This log.Warn is expected to be called rarely, so I'll keep this as is.

A bad example is a number returned in the form of string:

```
{
  "ReadingCelsius": "42"
}
```

This will cause a conversion error.
If this occurs in the real BMC, we should tolerate it with string-to-number conversion rather than returning an error.

Reply:
Dell's iDRAC returns "null" for Health, which caused an error here.
I updated a converter to tolerate "null" for Health.